### ISSUE TITLE
Incorrect behavior of `caller` instruction for dynamic calls

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/miden-vm#1478]
Currently, `dyncall` instruction translates into two operations in the VM: `call dyn`. This works mostly fine, but it does result in `caller` instruction returning an unintuitive (and probably wrong) value.

Specifically, the `caller` instruction should return the MAST root of the context we are currently in. So, if we do something like:
```
proc.foo
    caller
    # => [MAST_ROOT_OF_FOO, ...]
end

begin
    call.foo
end
```
But, if we use `dyncall` instruction, the result will be different:
```
proc.foo
    caller
    # => [HASH_DYN_NODE, ...]
end

begin
    procref.foo
    dyncall
end
```
This is because when we execute the `call` instruction, we set the MAST root of the new context to the hash of the callee. In case of `dyncall`, the callee is just a single `dyn` node - so, that's what gets used to set the MAST root.

Off the top of my head, there are two ways to fix this:
1. Update the semantics of the `call` operation to check if the next operation is `dyn`, and if so, use the top of the stack to set the MAST root. This may result in pretty complicated constraints - so, not sure if it will be worth it.
2. Introduce a new dedicate operation for `dyncall`. This will behave basically like `call` operation except it will use the top of the stack to set the MAST root of the context. I think `dyncall` will need to be a "very high degree" operation, and so we can do this only after we free up one of such operation slots (i.e., after [#1457](https://github.com/0xPolygonMiden/miden-vm/issues/1457)).

**Comment by plafer** [comment]
The `caller` instruction can only be called from within a syscall. And we currently cannot `call` or `dyncall` from within a syscall. Both of these programs even fail to assemble with error `Error:   x invalid instruction usage: 'caller' is only valid in kernel modules`. 

So unless I'm missing something, this inconsistency should never occur?

**Comment by bobbinth** [comment]
Ah yes, my example here was a bit sloppy. It should have been more like this:
```
kernel
---------
proc.foo
    caller
    # => [HASH_DYN_NODE, ...]
end

program
---------
proc.bar
    syscall.foo
end

begin
    procref.bar
    dyncall
end
```
The correct behavior here would be for the `caller` instruction to put the hash of `bar` onto the stack, but instead, right now it'll put the hash of the `DYN` node.

**Comment by plafer** [comment]
Closed by #1535 (also see [test](https://github.com/0xPolygonMiden/miden-vm/blob/ce26595583a41733bb31d8601ea89ce75fa98b7f/miden/tests/integration/flow_control/mod.rs#L428) here)
